### PR TITLE
Improve evolution dashboard filter descriptions

### LINKED ISSUE #290: Swift Evolution Dashboard filter status string improvements
This issue is a follow-on from PR #284 

Currently when the proposal status filter is turned on, a label appears next to it with the static text "Filtered By:" and a link that either lists the filtered proposal statuses or if there are too many of the chosen, the number of filters.

The upcoming filter flag filter added in PR #284 lists the filter status in the list header.

This issue is to:
1. Move the status string of the status/version filter to the list header
2. Make it more clear that the link next to the status filter button is used to edit the filter

The first part will put both filter status strings in a consistent place in the interface. It also gives the status/version filter more room to present its status string, allowing for more informative status.

The second part makes it more clear how to edit the status/version filter.

@kaishin @dempseyatgithub

### PR DESCRIPTION
**Pull request by dempseyatgithub**
This PR implements the changes discussed in Issue #290.

The high level changes are:
- Description of filtered statuses moves to the proposal list header, giving more space for content and consistent location with upcoming feature flag filter description.
- Link that shows/hides status filter panel now shows Show Filters/Hide Filters if no filters are selected, or the number of filters if any filter is selected.
- Upcoming feature flag filter description font weight is reduced to de-emphasize it a bit.

A test version of this PR is available at https://swift-test-site.xyz/swift-evolution

The specific changes are:
- Add proposal-list-header div and a status-filter-subhead div to handle multi-line proposal list header
- Remove now un-needed spans around filter-panel-toggle link
- Rename ‘toggle-filter-panel’ class to ‘filter-panel-toggle’ since it is a toggle control, not an action - Update JavaScript to target filter-panel-toggle directly - Consolidate filter-toggle css settings into filter-panel-toggle
- Separate flag-filter-description out from proposals-count-number span for different text styling
- Add css for new status-filter-subhead div, status-filter-description span, proposal-list-header div, flag-filter-description span
- Add media query for narrow screen size, such as iPad vertical split view. - Adjust first proposal list heading line to not wrap
- Hide and show status filter subhead in toggleStatusFiltering()
- Update _applyStatusFilter() to update status filter toggle text and subheading
- Update updateProposalsCount to set values of two different spans
- Replace updateFilterDescription() with updateStatusFilterToggleText()
- Add descriptionForSelectedStatuses() and listStringForStatuses() to generate status filter description
- Resolves #290

### CODE REVIEW

**Comment by dempseyatgithub** [comment]
@swift-ci test

**Comment by dempseyatgithub** [comment]
@swift-ci test

File: assets/javascripts/swift-evolution.js [hunk 1]
```diff
@@ -830,13 +835,24 @@ function _applyFlagFilter(matchingProposals) {
...
-  var selectedStates = [].map.call(allStateCheckboxes, function (checkbox) { return checkbox.value })
+  // get all checked state checkboxes, both status and version as an array
+  var allCheckedStateCheckboxes = Array.from(document.querySelectorAll('.filter-list input:checked'))
+  
+  // get checkbox values for all checked state checkboxes, both status and version
+  var selectedStates = allCheckedStateCheckboxes.map(function (checkbox) { return checkbox.value })
+
+  updateStatusFilterToggleText(selectedStates.length)
+
+  // get array of keys for only selected *statuses* to update the status filter subheading
```
[the remark this answers is not in the document]
Comment (dempseyatgithub) [reply assets/javascripts/swift-evolution.js]: This should be in the latest PR.

File: assets/javascripts/swift-evolution.js [hunk 2]
```diff
@@ -1171,3 +1175,51 @@ function cleanNumberFromState (state) {
 function addNumberToState (state, count) {
   return state + ' (' + count + ')'
 }
+
+/**
+* Generates the user-presentable description for an array of selected options. 
+* To prevent listing more than five statuses. When more than five statuses are selected
```
[the remark this answers is not in the document]
Comment (dempseyatgithub) [reply assets/javascripts/swift-evolution.js]: This should be in the latest PR.

File: assets/javascripts/swift-evolution.js [hunk 3]
```diff
@@ -1171,3 +1175,51 @@ function cleanNumberFromState (state) {
...
+* the generated string uses the form "All Statuses Except" followed by unselected statuses.
+*
+* @param {string[]} selectedOptions - each element is a key in the states objects. For example: '.accepted'.
+*/
+function descriptionForSelectedStatuses(selectedOptions) {
+  let allStateOptions = [
+      '.awaitingReview', '.scheduledForReview', '.activeReview', '.accepted',
+      '.previewing', '.implemented', '.returnedForRevision', '.rejected', '.withdrawn'
+   ]
+  let selectedCount  = selectedOptions.length
```
[the remark this answers is not in the document]
Comment (dempseyatgithub) [reply assets/javascripts/swift-evolution.js]: This should be in the latest PR.

File: assets/javascripts/swift-evolution.js [hunk 4]
```diff
@@ -830,13 +835,24 @@ function _applyFlagFilter(matchingProposals) {
  * @returns {Proposal[]} The results of applying the status filter.
  */
 function _applyStatusFilter(matchingProposals) {
-  // filter out proposals based on the grouping checkboxes
-  var allStateCheckboxes = document.querySelectorAll('.filter-list input:checked')
-  var selectedStates = [].map.call(allStateCheckboxes, function (checkbox) { return checkbox.value })
+  // get all checked state checkboxes, both status and version as an array
+  var allCheckedStateCheckboxes = Array.from(document.querySelectorAll('.filter-list input:checked'))
+  
+  // get checkbox values for all checked state checkboxes, both status and version
```
[the remark this answers is not in the document]
Comment (dempseyatgithub) [reply assets/javascripts/swift-evolution.js]: This should be in the latest PR

**Comment by dempseyatgithub** [comment]
I did one commit with all of the changes. I also made all inline comments in the file use sentence-case.

The latest commit also fixes an issue. If you turn on the status filter with no filters selected, the link text should change from "Show Filters" to "Hide Filters". Without this latest commit it does not - we needed to refresh that text after toggling the filter panel.

**Comment by dempseyatgithub** [comment]
@swift-ci test

**Comment by dempseyatgithub** [comment]
@swift-ci test

**Review by kaishin — APPROVED** [review]
I did a quick round of tests after the latest fixes and I didn't find any issues. :shipit:

**Pull request merged**